src/internal.cc

```diff
@@ -175,6 +175,9 @@ internalHostnameIs(const char *arg)
         if (0 == strcmp(arg, w->key))
             return 1;
 
+    if (0 == strcmp(arg, "localhost"))
+        return 1;
```

Comment:
The implied assumption that any Squid instance that receives a localhost URL listens on localhost is wrong AFAICT: An admin can configure Squid that does not listen on any IP addresses associated with the `localhost` name and send that Squid a request with a localhost URL.

How about checking for localhost here only when the relevant http_port does not specify an IP address or specifies an IP address that `localhost` resolves to? AFAICT, such a conditional check would take care of most common use cases without mishandling 

1. correct cache manager requests sent to a localhost Squid instance _through_ a non-localhost Squid instance
2. bad cache manager requests sent to a non-localhost Squid instance but targeting a localhost Squid instance

Reply:
> Alex: How about checking for localhost here only when the relevant http_port does not specify an IP address or specifies an IP address that localhost resolves to?

Moreover, the above logic can and should be extended to apply to _any_ requested domain name or IP that reach this function -- there is nothing specific about localhost here: If the requested target (including "localhost") matches[^1] the listening port address, return a truthy value.

[^1]: matches after the name resolution to get the targeted IP addresses (if such a resolution is necessary), of course

> Amos at https://bugs.squid-cache.org/show_bug.cgi?id=5283#c6:  Better checks will need to be implemented long-term.

I am not against merging this code change now if you think that is the best way forward, but we should add the corresponding comment. For example, "XXX: This check mishandles rare cases where this localhost-targeting request was not received on a localhost-associated port."

Reply:
Added the XXX in [07b4652](https://github.com/squid-cache/squid/pull/1406/commits/07b4652c0c715ed2ed4bfd24900cffd2c82d1013) since this should be fixed in v6.1 and the full fix will take longer and more testing.